### DISCUSSION TITLE
Injecting providers

### DISCUSSION THREAD

**Discussion by iammrsea** [adonisjs/core#1420] [Help] [answered]
Assuming I have an image service provider registered in AppProvider as follows:
```javascript
    this.$container.singleton("Adonis/Addons/ImageService", () => {
      const config = this.$container.use("Adonis/Core/Config").get("image");
      return new ImageService(config);
    });
```
If I try injecting it in my controller as:
```javascript
import imageServiceProvider from "@ioc:Adonis/Addons/ImageService";
```
typescript always yells at me because it can't resolve **@ioc:Adonis/Addons/ImageService**. I do get around the warning by doing this:
```javascript
//@ts-ignore
import imageServiceProvider from "@ioc:Adonis/Addons/ImageService";
```
And my question, is there a neater and better way of injecting my dependencies from the ioc container without adding **//@ts-ignore**?

**Comment by thetutlage** [comment, marked as answer]
Coz typescript has no idea that a module named `@ioc:Adonis/Addons/ImageService` exists. You have to define it using an [ambient module](https://www.typescriptlang.org/docs/handbook/modules.html#ambient-modules). 

Also check the source code of AdonisJS packages to see how ambient modules are defined

**Reply by msrumon** [reply]
And where do you recommend this module should be put?

**Reply by thetutlage** [reply]
Inside the contracts directory (if part of the application source code). For your packages you can keep them anywhere you like, but register the path inside package.json file. You can check typescript docs on publishing packages

**Reply by iammrsea** [reply]
Did some reading up on ambient modules and below is how I got it working for myself.
**created a root folder with index.d.ts** as shown in the snippet.
```javascript
//typings/index.d.ts
declare module "@ioc:Adonis/Addons/ImageService" {
//imports should be inside and not outside
  import File from "Contracts/file"; 
  import UploadedImage from "Contracts/uploadedImage";
  export default {
    uploadImage(file: File): Promise<UploadedImage>;,
    deleteImage(fileId: string): Promise<void>;,
  };
}
declare module 'another-module'{
}
```
And inside my controller, I have something like this:
```javascript
import imageProvider from "@ioc:Adonis/Addons/ImageService";
```
Everything works just fine. If you like you could place your declaration files inside **contracts** directory and it will also work
